File: token.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

type tokenKind int

const (
	unknown tokenKind = iota
	comment

	// language tokens
	comma
	dot
	dotdot
	colon
	leftParen
	rightParen
	leftBracket
	rightBracket
	leftBrace
	rightBrace
	assign // :=
	set    // =
	singleArrow
	doubleArrow

	// binary operators
	plus
	minus
	times
	divide
	modulus
	and
	or
	greater
	less
	eq
	geq
	leq
	neq

	// unary operators
	not
	negate

	// keywords
	breakKeyword
	ifKeyword
	elseKeyword
	forKeyword
	inKeyword
	returnKeyword

	// literals
	identifier
	trueLiteral
	falseLiteral
	stringLiteral
	numberLiteral
	nullLiteral
)

type position struct {
	line     int
	col      int
	filename string
}

func (p position) String() string {
	return fmt.Sprintf("[%d:%d]", p.line, p.col)
}

type token struct {
	kind    tokenKind
	pos     position
	payload string
}

func (t token) String() string {
	switch t.kind {
	case comma:
		return ","
	case dot:
		return "."
	case dotdot:
		return ".."
	case colon:
		return ":"
	case leftParen:
		return "("
	case rightParen:
		return ")"
	case leftBracket:
		return "["
	case rightBracket:
		return "]"
	case leftBrace:
		return "{"
	case rightBrace:
		return "}"
	case assign:
		return ":="
	case set:
		return "="
	case singleArrow:
		return "->"
	case doubleArrow:
		return "=>"

	case plus:
		return "+"
	case minus:
		return "-"
	case times:
		return "*"
	case divide:
		return "/"
	case modulus:
		return "%"
	case and:
		return "&&"
	case or:
		return "||"
	case greater:
		return ">"
	case less:
		return "<"
	case eq:
		return "=="
	case geq:
		return ">="
	case leq:
		return "<="
	case neq:
		return "!="

	case not:
		return "!"
	case negate:
		return "-"

	case breakKeyword:
		return "break"
	case ifKeyword:
		return "if"
	case elseKeyword:
		return "else"
	case forKeyword:
		return "for"
	case inKeyword:
		return "in"
	case returnKeyword:
		return "return"

	case identifier:
		return fmt.Sprintf("var(%s)", t.payload)
	case trueLiteral:
		return "true"
	case falseLiteral:
		return "false"
	case nullLiteral:
		return "null"
	case stringLiteral:
		return fmt.Sprintf("string(%s)", t.payload)
	case numberLiteral:
		return fmt.Sprintf("number(%s)", t.payload)

	default:
		return "<unknown>"
	}
}

type tokenizer struct {
	source   []rune
	index    int
	filename string
	line     int
	col      int
}

func newTokenizer(source string) tokenizer {
	return tokenizer{
		source:   []rune(source),
		index:    0,
		filename: "<input>",
		line:     1,
		col:      1,
	}
}

func (t *tokenizer) isEOF() bool {
	return t.index >= len(t.source)
}

func (t *tokenizer) next() rune {
	char := t.source[t.index]

	// ensure we don't go past EOF
	if t.index < len(t.source) {
		t.index++
	}

	if char == '\n' {
		t.line++
		t.col = 1
	} else {
		t.col++
	}

	return char
}

func (t *tokenizer) peek() rune {
	return t.source[t.index]
}

func (t *tokenizer) peekAhead(n int) rune {
	if t.index+n >= len(t.source) {
		return ' '
	}

	return t.source[t.index+n]
}

func (t *tokenizer) readUntil(ch rune) string {
	read := []rune{}
	for !t.isEOF() && t.peek() != ch {
		read = append(read, t.next())
	}

	return string(read)
}

func (t *tokenizer) readIdentifier() string {
	ident := []rune{}
	for !t.isEOF() {
		ch := t.peek()
		if unicode.IsLetter(ch) || unicode.IsDigit(ch) || ch == '_' || ch == '?' {
			ident = append(ident, t.next())
		} else {
			break
		}
	}

	return string(ident)
}

func (t *tokenizer) readNumber() string {
	literal := []rune{}
	dot := false

	for !t.isEOF() {
		ch := t.peek()
		if unicode.IsDigit(ch) {
			literal = append(literal, t.next())
		} else if ch == '.' && !dot {
			if t.peekAhead(1) == '.' {
				break
			}
			dot = true
			literal = append(literal, t.next())
		} else {
			break
		}
	}

	return string(literal)
}

func (t *tokenizer) pos() position {
	return position{
		line:     t.line,
		col:      t.col,
		filename: t.filename,
	}
}

func (t *tokenizer) nextToken() token {
	ch := t.next()

	switch ch {
	case ',':
		return token{kind: comma, pos: t.pos()}
	case '.':
		if !t.isEOF() && t.peek() == '.' {
			pos := t.pos()
			_ = t.next()

			return token{kind: dotdot, pos: pos}
		}
		return token{kind: dot, pos: t.pos()}
	case '(':
		return token{kind: leftParen, pos: t.pos()}
	case ')':
		return token{kind: rightParen, pos: t.pos()}
	case '[':
		return token{kind: leftBracket, pos: t.pos()}
	case ']':
		return token{kind: rightBracket, pos: t.pos()}
	case '{':
		return token{kind: leftBrace, pos: t.pos()}
	case '}':
		return token{kind: rightBrace, pos: t.pos()}
	case ':':
		if !t.isEOF() && t.peek() == '=' {
			pos := t.pos()
			t.next()
			return token{kind: assign, pos: pos}
		}
		return token{kind: colon, pos: t.pos()}

	case '=':
		if !t.isEOF() && t.peek() == '>' {
			pos := t.pos()
			t.next()
			return token{kind: doubleArrow, pos: pos}
		} else if !t.isEOF() && t.peek() == '=' {
			pos := t.pos()
			t.next()
			return token{kind: eq, pos: pos}
		}
		return token{kind: set, pos: t.pos()}
	case '>':
		if !t.isEOF() && t.peek() == '=' {
			pos := t.pos()
			t.next()
			return token{kind: geq, pos: pos}
		}
		return token{kind: greater, pos: t.pos()}
	case '<':
		if !t.isEOF() && t.peek() == '=' {
			pos := t.pos()
			t.next()
			return token{kind: leq, pos: pos}
		}
		return token{kind: less, pos: t.pos()}
	case '!':
		if !t.isEOF() && t.peek() == '=' {
			pos := t.pos()
			t.next()
			return token{kind: neq, pos: pos}
		}
		return token{kind: not, pos: t.pos()}

	case '+':
		return token{kind: plus, pos: t.pos()}
	case '-':
		if !t.isEOF() && t.peek() == '>' {
			pos := t.pos()
			t.next()
			return token{kind: singleArrow, pos: pos}
		}
		return token{kind: minus, pos: t.pos()}
	case '*':
		return token{kind: times, pos: t.pos()}
	case '/':
		if !t.isEOF() && t.peek() == '/' {
			pos := t.pos()
			t.next()
			commentString := strings.TrimSpace(t.readUntil('\n'))
			return token{kind: comment, pos: pos, payload: commentString}
		}
		return token{kind: divide, pos: t.pos()}
	case '%':
		return token{kind: modulus, pos: t.pos()}
	case '"':
		pos := t.pos()
		builder := strings.Builder{}
		for !t.isEOF() && t.peek() != '"' {
			ch := t.next()
			if ch == '\\' {
				builder.WriteRune(ch)
				if t.isEOF() {
					break
				} else {
					ch = t.next()
				}
			}
			builder.WriteRune(ch)
		}

		if t.isEOF() {
			return token{
				kind:    stringLiteral,
				pos:     pos,
				payload: builder.String(),
			}
		}

		t.next()
		return token{
			kind:    stringLiteral,
			pos:     pos,
			payload: builder.String(),
		}
	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		pos := t.pos()
		payload := string(ch) + t.readNumber()
		return token{kind: numberLiteral, pos: pos, payload: payload}
	default:
		pos := t.pos()
		payload := string(ch) + t.readIdentifier()
		switch payload {
		case "true":
			return token{kind: trueLiteral, pos: pos}
		case "false":
			return token{kind: falseLiteral, pos: pos}
		case "null":
			return token{kind: nullLiteral, pos: pos}
		case "if":
			return token{kind: ifKeyword, pos: pos}
		case "else":
			return token{kind: elseKeyword, pos: pos}
		case "for":
			return token{kind: forKeyword, pos: pos}
		case "in":
			return token{kind: inKeyword, pos: pos}
		case "break":
			return token{kind: breakKeyword, pos: pos}
		case "return":
			return token{kind: returnKeyword, pos: pos}
		default:
			return token{kind: identifier, pos: pos, payload: payload}
		}
	}
}

func (t *tokenizer) Tokenize() []token {
	tokens := []token{}

	// check for shebang and skip
	if !t.isEOF() && t.peek() == '#' && t.peekAhead(1) == '!' {
		t.readUntil('\n')
		if !t.isEOF() {
			t.next()
		}
	}

	// skip whitespace
	for !t.isEOF() && unicode.IsSpace(t.peek()) {
		t.next()
	}

	last := token{kind: comma}

	for !t.isEOF() {
		next := t.nextToken()

		// separate expressions
		if (last.kind != leftParen && last.kind != leftBracket &&
			last.kind != leftBrace && last.kind != comma) &&
			(next.kind == rightParen || next.kind == rightBracket ||
				next.kind == rightBrace) {
			tokens = append(tokens, token{
				kind: comma,
				pos:  t.pos(),
			})
		}

		if next.kind == comment {
			next = last
		} else {
			tokens = append(tokens, next)
		}

		for !t.isEOF() && unicode.IsSpace(t.peek()) {
			if t.peek() == '\n' {
				// if we hit a token that can end a statement, insert a comma
				switch next.kind {
				case rightParen, rightBrace, rightBracket, identifier, numberLiteral, stringLiteral, trueLiteral, falseLiteral, nullLiteral:
					next = token{
						kind: comma,
						pos:  t.pos(),
					}
					tokens = append(tokens, next)
				}
			}
			t.next()
		}

		if next.kind != comment {
			last = next
		}
	}

	if last.kind != comma {
		tokens = append(tokens, token{
			kind: comma,
			pos:  t.pos(),
		})
	}

	return tokens
}

File: src/token.zig
const std = @import("std");

const keywords = std.ComptimeStringMap(Kind, .{
    .{ "break", .breakKeyword },
    .{ "continue", .continueKeyword },
    .{ "if", .ifKeyword },
    .{ "else", .elseKeyword },
    .{ "for", .forKeyword },
    .{ "in", .inKeyword },
    .{ "return", .returnKeyword },

    .{ "true", .trueLiteral },
    .{ "false", .falseLiteral },
    .{ "null", .nullLiteral },
});

pub const Kind = enum {
    comment,

    // language tokens
    comma,
    dot,
    dotdot,
    colon,
    leftParen,
    rightParen,
    leftBracket,
    rightBracket,
    leftBrace,
    rightBrace,
    assign, // :=
    set, // =
    singleArrow,
    doubleArrow,

    // binary operators
    plus,
    minus,
    times,
    divide,
    modulus,
    @"and",
    @"or",
    greater,
    less,
    eq,
    geq,
    leq,
    neq,

    // unary operators
    not,
    negate,

    // keywords
    breakKeyword,
    continueKeyword,
    ifKeyword,
    elseKeyword,
    forKeyword,
    inKeyword,
    returnKeyword,

    // literals
    identifier,
    trueLiteral,
    falseLiteral,
    stringLiteral,
    numberLiteral,
    nullLiteral,
};

const Position = struct {
    line: u32,
    column: u32,
    filename: []const u8,

    pub fn format(
        self: Position,
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt;
        _ = options;
        try writer.print("[{}:{}]", self.line, self.column);
    }
};

pub const Token = struct {
    kind: Kind,
    pos: Position,
    payload: ?[]const u8 = null,

    pub fn string(self: Token) []const u8 {
        return switch (self.kind) {
            .comma => ",",
            .dot => ".",
            .dotdot => "..",
            .colon => ":",
            .leftParen => "(",
            .rightParen => ")",
            .leftBracket => "[",
            .rightBracket => "]",
            .leftBrace => "{",
            .rightBrace => "}",
            .assign => ":=",
            .set => "=",
            .singleArrow => "->",
            .doubleArrow => "=>",

            .plus => "+",
            .minus => "-",
            .times => "*",
            .divide => "/",
            .modulus => "%",
            .@"and" => "and",
            .@"or" => "or",
            .greater => ">",
            .less => "<",
            .eq => "==",
            .geq => ">=",
            .leq => "<=",
            .neq => "!=",

            .not => "!",
            .negate => "-",

            .breakKeyword => "break",
            .continueKeyword => "continue",
            .ifKeyword => "if",
            .elseKeyword => "else",
            .forKeyword => "for",
            .inKeyword => "in",
            .returnKeyword => "return",

            .identifier => self.payload.?,
            .trueLiteral => "true",
            .falseLiteral => "false",
            .stringLiteral => self.payload.?,
            .numberLiteral => self.payload.?,
            .nullLiteral => "null",

            .comment => "comment",
        };
    }

    pub fn format(
        self: Token,
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt;
        _ = options;
        switch (self.kind) {
            .identifier => try writer.print("var({s})", .{self.payload.?}),
            .stringLiteral => try writer.print("string({s})", .{self.payload.?}),
            .numberLiteral => try writer.print("number({s})", .{self.payload.?}),
            else => try writer.print("{s}", .{self.string()}),
        }
    }
};

pub const Tokenizer = struct {
    const This = @This();
    source: []const u8,
    filename: []const u8,

    index: usize,
    line: u32,
    column: u32,

    pub fn init(source: []const u8, filename: ?[]const u8) Tokenizer {
        return Tokenizer{
            .source = source,
            .filename = if (filename) |f| f else "unknown",
            .index = 0,
            .line = 0,
            .column = 0,
        };
    }

    fn isEOF(t: *This) bool {
        return t.index >= t.source.len;
    }

    fn next(t: *Tokenizer) u8 {
        var char = t.source[t.index];

        // ensure we don't go past EOF
        if (t.index < t.source.len) {
            t.index += 1;
        }

        if (char == '\n') {
            t.line += 1;
            t.column = 0;
        } else {
            t.column += 1;
        }

        return char;
    }

    fn peek(t: *Tokenizer) u8 {
        return t.source[t.index];
    }

    fn peekAhead(t: *Tokenizer, n: usize) ?u8 {
        if (t.index + n < t.source.len) {
            return t.source[t.index + n];
        } else {
            return null;
        }
    }

    fn readUntil(t: *Tokenizer, ch: u8) []const u8 {
        var start = t.index - 1;
        while (!t.isEOF() and t.peek() != ch) {
            _ = t.next();
        }
        return t.source[start..t.index];
    }

    fn readIdentifier(t: *Tokenizer) []const u8 {
        var start = t.index - 1;
        while (!t.isEOF()) {
            var ch = t.peek();
            if (ch == '_' or ch == '?' or isLetter(ch) or isDigit(ch)) {
                _ = t.next();
            } else {
                break;
            }
        }

        return t.source[start..t.index];
    }

    fn readNumber(t: *Tokenizer) []const u8 {
        var start = t.index - 1;
        var dot = false;
        while (!t.isEOF()) {
            var ch = t.peek();
            if (isDigit(ch)) {
                _ = t.next();
            } else if (ch == '.' and !dot) {
                // range operator
                if (t.peekAhead(1) == '.') {
                    break;
                }

                dot = true;
                _ = t.next();
            } else {
                break;
            }
        }

        return t.source[start..t.index];
    }

    fn position(t: *Tokenizer) Position {
        return Position{
            .line = t.line,
            .column = t.column,
            .filename = t.filename,
        };
    }

    fn branch(t: *Tokenizer, cond: u8, then: Kind, otherwise: Kind) Token {
        var pos = t.position();

        if (!t.isEOF() and t.peek() == cond) {
            _ = t.next();
            return Token{ .kind = then, .pos = pos };
        } else {
            return Token{ .kind = otherwise, .pos = pos };
        }
    }

    fn nextToken(t: *Tokenizer) Token {
        var ch = t.next();

        switch (ch) {
            ',' => return Token{ .kind = .comma, .pos = t.position() },
            '.' => {
                if (!t.isEOF() and t.peek() == '.') {
                    var pos = t.position();
                    _ = t.next();
                    return Token{ .kind = .dotdot, .pos = pos };
                } else {
                    return Token{ .kind = .dot, .pos = t.position() };
                }
            },
            '(' => return Token{ .kind = .leftParen, .pos = t.position() },
            ')' => return Token{ .kind = .rightParen, .pos = t.position() },
            '[' => return Token{ .kind = .leftBracket, .pos = t.position() },
            ']' => return Token{ .kind = .rightBracket, .pos = t.position() },
            '{' => return Token{ .kind = .leftBrace, .pos = t.position() },
            '}' => return Token{ .kind = .rightBrace, .pos = t.position() },

            ':' => return t.branch('=', .assign, .colon),
            '>' => return t.branch('=', .geq, .greater),
            '<' => return t.branch('=', .leq, .less),
            '!' => return t.branch('=', .neq, .not),
            '=' => {
                var pos = t.position();
                if (!t.isEOF() and t.peek() == '>') {
                    _ = t.next();
                    return Token{ .kind = .doubleArrow, .pos = pos };
                } else if (!t.isEOF() and t.peek() == '=') {
                    _ = t.next();
                    return Token{ .kind = .eq, .pos = pos };
                } else {
                    return Token{ .kind = .set, .pos = pos };
                }
            },

            '+' => return Token{ .kind = .plus, .pos = t.position() },
            '-' => return t.branch('>', .singleArrow, .minus),
            '*' => return Token{ .kind = .times, .pos = t.position() },
            '/' => return Token{ .kind = .divide, .pos = t.position() },
            '%' => return Token{ .kind = .modulus, .pos = t.position() },

            '"' => {
                var pos = t.position();
                var start = t.index - 1;

                while (!t.isEOF() and t.peek() != '"') {
                    var char = t.next();
                    if (char == '"') {
                        break;
                    } else if (ch == '\\') {
                        if (t.isEOF()) {
                            break;
                        }
                        _ = t.next();
                    }
                }

                if (t.isEOF()) {
                    return Token{ .kind = .stringLiteral, .pos = pos, .payload = t.source[start + 1 .. t.index] };
                } else {
                    _ = t.next();
                    return Token{ .kind = .stringLiteral, .pos = pos, .payload = t.source[start + 1 .. t.index - 1] };
                }
            },
            '0'...'9' => {
                var pos = t.position();
                var payload = t.readNumber();
                return Token{ .kind = .numberLiteral, .pos = pos, .payload = payload };
            },
            else => {
                var pos = t.position();
                var payload = t.readIdentifier();

                if (keywords.get(payload)) |keyword| {
                    return Token{ .kind = keyword, .pos = pos };
                }

                return Token{ .kind = .identifier, .pos = pos, .payload = payload };
            },
        }
    }

    pub fn tokenize(t: *Tokenizer, tokens: *std.ArrayList(Token)) !void {
        // check for shebang and skip it
        if (!t.isEOF() and t.peek() == '#' and t.peekAhead(1) == '!') {
            _ = t.readUntil('\n');
            if (!t.isEOF()) {
                _ = t.next();
            }
        }

        // skip whitespace
        while (!t.isEOF() and isSpace(t.peek())) {
            _ = t.next();
        }

        var last = Token{ .kind = .comma, .pos = t.position() };

        while (!t.isEOF()) {
            var ch = t.nextToken();

            // separate expressions
            if ((last.kind != .leftParen and last.kind != .leftBracket and last.kind != .leftBrace and last.kind != .comma) and (ch.kind == .identifier or ch.kind == .rightBracket or ch.kind == .rightBrace)) {
                try tokens.append(Token{ .kind = .comma, .pos = t.position() });
            }

            if (ch.kind == .comment) {
                ch = last;
            } else {
                try tokens.append(ch);
            }

            while (!t.isEOF() and isSpace(t.peek())) {
                if (t.peek() == '\n') {
                    switch (ch.kind) {
                        // if we hit a toke nthat can end a statement, insert a comma
                        .rightParen, .rightBrace, .rightBracket, .identifier, .numberLiteral, .stringLiteral, .trueLiteral, .falseLiteral, .nullLiteral => {
                            try tokens.append(Token{ .kind = .comma, .pos = t.position() });
                        },
                        else => {},
                    }
                }

                _ = t.next();
            }

            if (ch.kind != .comment) {
                last = ch;
            }
        }

        if (last.kind != .comma) {
            try tokens.append(Token{ .kind = .comma, .pos = t.position() });
        }
    }
};

fn isLetter(ch: u8) bool {
    return (ch >= 'a' and ch <= 'z') or (ch >= 'A' and ch <= 'Z');
}

fn isDigit(ch: u8) bool {
    return ch >= '0' and ch <= '9';
}

fn isSpace(ch: u8) bool {
    return ch == ' ' or ch == '\t' or ch == '\n' or ch == '\r';
}
